processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -275,92 +274,111 @@ public void close()
-    private volatile String[] sortedVals = null;
-    final ConcurrentMap<String, String> poorMansInterning = Maps.newConcurrentMap();
+    private final String dimension;
+    private final ConcurrentMap<String, Integer> valueToId = Maps.newConcurrentMap();
 
-    public OnHeapDimDim()
+    // for faster access
+    @GuardedBy("valueToId")
+    private String[] idToValue = new String[32];
+
+    public OnHeapDimDim(String dimension)
     {
-      BiMap<String, Integer> biMap = Maps.synchronizedBiMap(HashBiMap.<String, Integer>create());
-      falseIds = biMap;
-      falseIdsReverse = biMap.inverse();
+      this.dimension = dimension;
     }
 
-    /**
-     * Returns the interned String value to allow fast comparisons using `==` instead of `.equals()`
-     *
-     * @see io.druid.segment.incremental.IncrementalIndexStorageAdapter.EntryHolderValueMatcherFactory#makeValueMatcher(String, String)
-     */
-    public String get(String str)
+    @Override
+    public String dimension()
     {
-      String prev = poorMansInterning.putIfAbsent(str, str);
-      return prev != null ? prev : str;
+      return dimension;
     }
 
     public int getId(String value)
     {
-      final Integer id = falseIds.get(value);
+      final Integer id = valueToId.get(value);
       return id == null ? -1 : id;
     }
 
     public String getValue(int id)
     {
-      return falseIdsReverse.get(id);
+      synchronized (valueToId) {
+        return idToValue[id];
+      }
     }
 
     public boolean contains(String value)
     {
-      return falseIds.containsKey(value);
+      return valueToId.containsKey(value);
     }
 
     public int size()
     {
-      return falseIds.size();
+      return valueToId.size();
     }
 
-    public synchronized int add(String value)
+    public int add(String value)
     {
-      int id = falseIds.size();
-      falseIds.put(value, id);
-      return id;
+      int index = size();
+      Integer prev = valueToId.putIfAbsent(value, index);
+      if (prev != null) {
+        return prev;
+      }
+      addNewValue(index, value);
+      return index;
     }
 
-    public int getSortedId(String value)
+    private void addNewValue(int index, String value)
     {
-      assertSorted();
-      return Arrays.binarySearch(sortedVals, value);
+      synchronized (valueToId) {
+        if (index >= idToValue.length) {
+          idToValue = Arrays.copyOf(idToValue, idToValue.length << 1);
+        }
+        idToValue[index] = value;
+      }
     }
 
-    public String getSortedValue(int index)
+    public OnHeapDimLookup sort()
     {
-      assertSorted();
-      return sortedVals[index];
+      synchronized (valueToId) {
+        return new OnHeapDimLookup(idToValue, size());
+      }
     }
+  }
 
-    public void sort()
-    {
-      if (sortedVals == null) {
-        sortedVals = new String[falseIds.size()];
+  private static class OnHeapDimLookup implements DimLookup
```

Comment:
Can you change DimLookup class name to something that indicates it's used for sorted values, e.g. SortedDimLookup